pkg/ccl/changefeedccl/event_processing.go

```diff
@@ -240,3 +330,238 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+	// parallelEventConsumer.Close() is called.
+	doneCh chan struct{}
+
+	// makeConsumer creates a single-threaded consumer
+	// which encodes and emits events.
+	makeConsumer func() (eventConsumer, error)
+
+	// numWorkers is the number of worker threads.
+	numWorkers int64
+	// workerCh stores the event buffer for each worker.
+	workerCh []chan kvevent.Event
+	// workerChSize is the maximum number of events a worker can buffer.
+	workerChSize int64
+
+	// spanFrontier stores the frontier for the aggregator
+	// that spawned this event consumer.
+	spanFrontier *span.Frontier
+
+	// termErr and termCh are used to save the first error that occurs
+	// in any worker and signal all workers to stop.
+	//
+	// inFlight, waiting, and flushCh are used to flush the sink.
+	//
+	// flushFrontier tracks the local frontier. It is set to the
+	// spanFrontier upon flushing. This guarantees that workers are
+	// not buffering events which have timestamps lower than the frontier.
+	mu struct {
+		syncutil.Mutex
+		termErr error
+
+		inFlight      int
+		waiting       bool
+		flushFrontier hlc.Timestamp
+	}
+	termCh  chan struct{}
+	flushCh chan struct{}
+}
+
+var _ eventConsumer = (*parallelEventConsumer)(nil)
+
+func (c *parallelEventConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Event) error {
+	startTime := timeutil.Now().UnixNano()
+	defer func() {
+		time := timeutil.Now().UnixNano()
+		c.metrics.ParallelConsumerConsumeNanos.Inc(time - startTime)
+	}()
+
+	bucket := c.getBucketForEvent(ev)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.termCh:
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		return c.mu.termErr
+	case c.workerCh[bucket] <- ev:
+		c.incInFlight()
+		return nil
+	}
+}
+
+// getBucketForEvent returns a deterministic value  between [0,
+// parallelEventConsumer.numWorkers).
+//
+// This is used to ensure events of the same key get sent to the same worker.
+// Events of the same key are sent to the same worker so per-key ordering is
+// maintained.
+func (c *parallelEventConsumer) getBucketForEvent(ev kvevent.Event) int64 {
+	c.hash.Reset()
+	c.hash.Write(ev.KV().Key)
+	return int64(c.hash.Sum32()) % c.numWorkers
+}
+
+func (c *parallelEventConsumer) startWorkers() error {
+	// Create the consumers in a separate loop so that returning
+	// in case of an error is simple and does not require
+	// shutting down goroutines.
+	consumers := make([]eventConsumer, c.numWorkers)
+	for i := int64(0); i < c.numWorkers; i++ {
+		consumer, err := c.makeConsumer()
+		if err != nil {
+			return err
+		}
+		consumers[i] = consumer
+	}
+
+	for i := int64(0); i < c.numWorkers; i++ {
+		c.workerCh[i] = make(chan kvevent.Event, c.workerChSize)
+
+		// c.g.GoCtx requires a func(context.Context), so
+		// a closure is used to pass the consumer and id to
+		// the worker.
+		id := i
+		consumer := consumers[i]
+		workerClosure := func(ctx2 context.Context) error {
+			return c.workerLoop(ctx2, consumer, id)
+		}
+		c.g.GoCtx(workerClosure)
```

Comment:
have you seen c.g.GroupWorkers utility function?

Reply:
That utility function (and a few others) block until the worker goroutines return. We want them to run async while we return on the main thread here.